perf(linter/no-inner-declarations): remove unnecessary code and reduce branches

overlookmotel opened the pull request:
1. Remove unnecessary check that grandparent of node is a `Function` or `ArrowFunctionExpression` when its parent is a `FunctionBody`. `FunctionBody` is always the child of `Function` or `ArrowFunctionExpression`, so this check is redundant.

2. If node is a `VariableStatement` or `Function`, it always has a parent node. So `unwrap` the parent instead of branching `if let Some(parent_node) = parent`. It's always `Some`.

The 2nd also allows skipping 1 turn of the loop further down because:

* If direct parent is a `StaticBlock`, we already exited.
* It's impossible for a variable or function declaration's parent to be a `Function` - there'd have to be a `FunctionBody` in between.

overlookmotel commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/11633?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#11633** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/11633?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/11633?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#11632** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/11632?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=11633).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

camc314 commented:
### Merge activity

* **Jun 12, 11:29 AM UTC**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

overlookmotel committed:
perf(linter/no-inner-declarations): remove unnecessary code and reduce branches (#11633)

1. Remove unnecessary check that grandparent of node is a `Function` or `ArrowFunctionExpression` when its parent is a `FunctionBody`. `FunctionBody` is always the child of `Function` or `ArrowFunctionExpression`, so this check is redundant.

2. If node is a `VariableStatement` or `Function`, it always has a parent node. So `unwrap` the parent instead of branching `if let Some(parent_node) = parent`. It's always `Some`.

The 2nd also allows skipping 1 turn of the loop further down because:

* If direct parent is a `StaticBlock`, we already exited.
* It's impossible for a variable or function declaration's parent to be a `Function` - there'd have to be a `FunctionBody` in between.
